lib/model/settings.dart

```diff
@@ -24,3 +28,44 @@ enum ThemeSetting {
     };
   }
 }
+
+/// What browser the user has set to use for opening links in messages.
+///
+/// See https://chat.zulip.org/#narrow/stream/48-mobile/topic/in-app.20browser
+/// for the reasoning behind these options.
+///
+/// Renaming existing enum values will invalidate the database.
+/// Write a migration if such a change is necessary.
+enum BrowserPreference {
+  /// Use the in-app browser for HTTP links.
+  ///
+  /// For other types of links (e.g. mailto), this translates to
+  /// [UrlLaunchMode.platformDefault], so that non-in-app browsers/apps can
+  /// still be used.
+  inApp,
+
+  /// Use the user's default browser app.
+  external,
+}
+
+extension GlobalSettingsHelpers on GlobalSettingsData {
+  BrowserPreference get effectiveBrowserPreference {
+    if (browserPreference != null) return browserPreference!;
+    return switch (defaultTargetPlatform) {
+      // On iOS we prefer UrlLaunchMode.externalApplication because (for
+      // HTTP URLs) UrlLaunchMode.platformDefault uses SFSafariViewController,
+      // which gives an awkward UX as described here:
+      //  https://chat.zulip.org/#narrow/stream/48-mobile/topic/in-app.20browser/near/1169118
+      TargetPlatform.iOS => BrowserPreference.external,
+      _ => BrowserPreference.inApp,
+    };
+  }
+
+  UrlLaunchMode getUrlLaunchMode(Uri url) {
+    return switch (effectiveBrowserPreference) {
+      BrowserPreference.inApp => url.scheme == 'http' || url.scheme == 'https'
+        ? UrlLaunchMode.inAppBrowserView : UrlLaunchMode.platformDefault,
```

Comment:
This condition calls for a comment to explain why we need this (and why it takes the particular shape it does).

Much of the needed information is in the commit message, so just needs to be moved from there. I also had some changes I wanted to make to the framing of that, so I made a revision which I'll push along with this review.

Reply:
Ah I see! It makes sense that the upstream behavior needs to change.